### PR TITLE
Add r2d1 agents

### PR DESCRIPTION
**Pull request by jinPrelude**
1. Add r2d1-style recurrent agents combined with DQN, C51, and IQN.
2. Add total_step & total_test_step to wandb log.
3. Change total_step to start counting after warmup.
4. Change test startpoint after warmup

### CODE REVIEW

File: rl_algorithms/common/networks/brain.py [hunk 1]
```diff
@@ -54,3 +58,142 @@ def calculate_fc_input_size(self, state_dim: tuple):
...
+                + 1,  # 1 is for prev_reward
+                head_cfg.configs.rnn_hidden_size,
+                batch_first=True,
+            )
+        else:
+            self.backbone = build_backbone(backbone_cfg)
+            head_cfg.configs.input_size = self.calculate_fc_input_size(
+                head_cfg.configs.state_size
+            )
+            self.fc = nn.Linear(
```
Comment (MrSyee) [inline rl_algorithms/common/networks/brain.py:90]: remove duplicate code

File: rl_algorithms/common/abstract/agent.py [hunk 2]
```diff
@@ -151,7 +152,7 @@ def _test(self, interim_test: bool = False):
             test_num = self.args.interim_test_num
         else:
             test_num = self.args.episode_num
-
+        score_list = []
```
Comment (MrSyee) [inline rl_algorithms/common/abstract/agent.py:155]: should put line break before this line

File: rl_algorithms/common/buffer/recurrent_replay_buffer.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
+# -*- coding: utf-8 -*-
```
Comment (MrSyee) [inline rl_algorithms/common/buffer/recurrent_replay_buffer.py:1]: should change wrapper class

File: rl_algorithms/common/helper_functions.py [hunk 4]
```diff
@@ -88,3 +88,67 @@ def get_n_step_info(
         next_state, done = (n_s, d) if d else (next_state, done)
 
     return reward, next_state, done
+
+
+def infer_leading_dims(tensor, dim):
```
Comment (MrSyee) [inline rl_algorithms/common/helper_functions.py:93]: should add type annotation

Comment (MrSyee) [inline rl_algorithms/common/helper_functions.py:93]: I think these methods don't use for common. Should make `recurrent/utils.py` and move them.

File: rl_algorithms/common/helper_functions.py [hunk 5]
```diff
@@ -88,3 +88,67 @@ def get_n_step_info(
...
+    tensors = tensors if is_seq else (tensors,)
+    if lead_dim == 2:  # (Put T dim.)
+        tensors = tuple(t.view((T, B) + t.shape[1:]) for t in tensors)
+    if lead_dim == 0:  # (Remove B=1 dim.)
+        assert B == 1
+        tensors = tuple(t.squeeze(0) for t in tensors)
+    return tensors if is_seq else tensors[0]
+
+
+def valid_from_done(done):
```
Comment (MrSyee) [inline rl_algorithms/common/helper_functions.py:140]: It is difficult to understand this function's role though I see this function name and contents. Should modify function or function name.

File: rl_algorithms/common/networks/brain.py [hunk 6]
```diff
@@ -54,3 +58,142 @@ def calculate_fc_input_size(self, state_dim: tuple):
...
+        self,
+        x: torch.Tensor,
+        hidden: torch.Tensor,
+        prev_action: torch.Tensor,
+        prev_reward: torch.Tensor,
+    ):
+        """Forward method implementation. Use in get_action method in agent."""
+        if isinstance(self.backbone, nn.Module):
+            x = x / 255.0
+            lead_dim, T, B, x_shape = infer_leading_dims(x, 3)
```
Comment (MrSyee) [inline rl_algorithms/common/networks/brain.py:114]: should change variables name
e.g.
```
T -> seq_len
B -> batch_len
```

File: rl_algorithms/common/networks/brain.py [hunk 7]
```diff
@@ -54,3 +58,142 @@ def calculate_fc_input_size(self, state_dim: tuple):
...
+        return q, hidden
+
+    def forward_(
+        self,
+        x: torch.Tensor,
+        hidden: torch.Tensor,
+        prev_action: torch.Tensor,
+        prev_reward: torch.Tensor,
+        n_tau_samples: int = None,
+    ):
```
Comment (MrSyee) [inline rl_algorithms/common/networks/brain.py:151]: should add reference why it takes for prev_action, prev_reward
